classes/04class/iac/README.md

```diff
@@ -0,0 +1,304 @@
+
+#### Cost
+Repeatable tasks that are transformed into code can reduce costs. Instead of spending 2 hours creating an environment manually that needs to be created weekly, though code, that environment can take less than 5 minutes. 
+
+With that, you're not only reducing the cost to deliver that environment, but the engineer is providing it will be able to focus on more critical things for the company.
+
+#### Speed
+IaC will make the resources available faster, enabling developers to start working sooner. Especially for environments with several changes or with fast innovation pace, where new environments need to be created frequently, IaC can enable this fast pace.
+
+#### Risk
+Because everything is defined through code, you can ensure that no manual mistake can be made during any of the configuration. Also, you'll know what is running on your environment, considering that deployments are only made through code, and no manual changes are possible.
+
+Decreasing the possibility of manual mistakes, you increase the availability and reliability of your environment and reduce your application downtime.
+
+### IaC Models
+There are multiple models of IaC, and the two main models are Imperative and Declarative. The more natural way to understand the difference between those two models would be that declarative focus on what and imperative focus on how.
+
+In other words, in the declarative model, you define what the desired state is, and the system will execute the necessary steps to achieve that state. In the imperative model, you need to determine what actions need to be performed as well as the order, so the machine knows what it needs to do to achieve the desired state.
+
+### IaC Methods
+There are two methods in IaC. Push and Pull. The difference is who starts the communication between the tool and the resource.
+
+In the pull method, the resource being managed by the IaC tool will pull the tool to check if there is any change in the desired state. If any new state definition is found in the IaC tool, the resource will converge to the desired state(either through imperative or declarative models, depending on the tool).
+
+In the push method, the resource being managed stays in a `sleep mode`, and the communication is initiated by the IaC tool. In that case, the tool will push the desired state to the resource being managed and the resource will converge to the desired state (again, either through imperative or declarative model).
+
+## Terraform Foundation
+[Terraform](https://www.terraform.io/) is one of the main IaC tools available, and it uses a Decalative model with a push method. It can be used to manage resources in multiple cloud providers(including AWS, Azure and GCP) as well as manage resources in on-premisses(like in VMware and Openshift).
+
+It's a powerful tool that can be used to create, maintain and delete entire IT environments. During this material, we'll focus on how to use Terraform to manage AWS resources.
```

Comment:
This powerful tool can be used to create, maintain... This material will focus on how to use..

Reply:
fixed